File: Find_All_Anagrams_in_a_String/solution2.py
from collections import Counter

class Solution:
  def findAnagrams(self, s: str, p: str) -> List[int]:
    ans = []
    if s == None or p == None: return ans
    
    sLen = len(s)
    pLen = len(p)
    if sLen == 0 or pLen == 0 or sLen < pLen: return ans
    
    b = e = 0
    lackCount = pLen
    lackMap = Counter(p)
    
    c = ''
    while e < pLen:
      c = s[e]
      lackMap[c] -= 1
      if lackMap[c] >= 0: lackCount -= 1
      e += 1
    if lackCount == 0: ans.append(b)
    
    while e < sLen:
      c = s[b]
      lackMap[c] += 1
      if lackMap[c] > 0: lackCount += 1
      b += 1
      
      c = s[e]
      lackMap[c] -= 1
      if lackMap[c] >= 0: lackCount -= 1
      e += 1
      
      if lackCount == 0: ans.append(b)
        
    return ans
    

File: Find_All_Anagrams_in_a_String/solution2.java
class Solution {
  public List<Integer> findAnagrams(String s, String p) {
    ArrayList<Integer> ans = new ArrayList<>();
    
    if (s == null || p == null) return ans;
    
    int sLen = s.length(), pLen = p.length();
    if (sLen == 0 || pLen == 0 || sLen < pLen) return ans;
    
    int b = 0, e = 0, lackCount = pLen;
    int[] lackMap = new int[26];
    for (char c : p.toCharArray()) {
      lackMap[c - 'a']++;
    }
    
    int pos;
    while (e < pLen) {
      pos = s.charAt(e) - 'a';
      lackMap[pos]--;
      if (lackMap[pos] >= 0) lackCount--;
      e++;
    }
    if (lackCount == 0) ans.add(b);
    
    while (e < sLen) {
      pos = s.charAt(b) - 'a';
      lackMap[pos]++;
      if (lackMap[pos] > 0) lackCount++;
      b++;
      
      pos = s.charAt(e) - 'a';
      lackMap[pos]--;
      if (lackMap[pos] >= 0) lackCount--;
      e++;
      
      if (lackCount == 0) ans.add(b);
    }
    
    return ans;
  }
}
